src/Duration.php

```diff
@@ -0,0 +1,187 @@
+            return self::fromPeriodCollection($value);
+        }
+
+        if ($value instanceof DateInterval) {
+            return self::fromDateInterval($value);
+        }
+
+        $stringValue = (string) $value;
+
+        if (0 === strpos($stringValue, 'P')) {
+            return self::fromDateInterval(new DateInterval($stringValue));
+        }
+
+        return self::fromDateInterval(DateInterval::createFromDateString($stringValue));
+    }
+
+    public static function fromPeriod(Period $period): self
+    {
+        $start = $period->getIncludedStart();
+        $end = $period->getIncludedEnd();
+        $precision = $period->getPrecisionMask();
+
+        if ($precision === Precision::SECOND) {
+            $end = $end->add(new DateInterval('PT1S'));
+        } elseif ($precision === Precision::MINUTE) {
+            $end = $end->add(new DateInterval('PT1M'));
+        } elseif ($precision === Precision::HOUR) {
+            $end = $end->add(new DateInterval('PT1H'));
+        } elseif ($precision === Precision::DAY) {
+            $end = $end->add(new DateInterval('PT24H'));
+        } else {
+            throw new \InvalidArgumentException('A duration can only be determined up to a precision of Precision::DAY');
+        }
+
+        $duration = new self();
+        $duration->seconds = $end->getTimestamp() - $start->getTimestamp();
+        $duration->precision = $precision;
+
+        return $duration;
+    }
+
+    public static function fromPeriodCollection(PeriodCollection $collection): self
+    {
+        $duration = self::none();
+
+        foreach ($collection as $period) {
+            $duration = $duration->withAdded($period->duration());
+        }
+
+        return $duration;
+    }
+
+    public static function fromDateInterval(DateInterval $interval): self
+    {
+        $now = new DateTimeImmutable();
+        $then = $now->add($interval);
+
+        $duration = new self();
+        $duration->seconds = $then->getTimestamp() - $now->getTimestamp();
+        $duration->precision = self::determinePrecisionFromSeconds($duration->seconds);
+
+        return $duration;
+    }
+
+    public static function none(): self
+    {
+        $new = new self();
+        $new->seconds = 0;
+        $new->precision = Precision::SECOND;
+
+        return $new;
+    }
+
+    public function length(int $precision): int
+    {
+        switch ($precision) {
+            case Precision::SECOND:
+                return $this->seconds;
+            case Precision::MINUTE:
+                return intdiv($this->seconds, 60);
+            case Precision::HOUR:
+                return intdiv($this->seconds, 3600);
+            case Precision::DAY:
+                return intdiv($this->seconds, 86400);
+            default:
+                throw new \InvalidArgumentException('Unsupported precision for durations');
+        }
+    }
+
+    public function precision(): int
+    {
+        return $this->precision;
```

Comment:
Now I'm wondering whether we need some abstraction for precision. I really like the binary flag for internal usages, but they are not that user-friendly to the outside. What do you think?

PS: I'm not saying we should change this PR, it's just a remark reading this.

Reply:
I'm totally for changing the PR :D. How would you feel about adding a dependency to the package? Don't know if you already heard of it, but I recently stumbled over a package called [spatie/enum](https://github.com/spatie/enum) which looks quite promising :trollface:

Reply:
Let me think about that, anyways we'll discuss it apart from this PR :)